### PR TITLE
Add gpu call optimisations

### PR DESCRIPTION
**Pull request by maroroman**
This PR adds gpu optimisations made during the creation of this PR: #573 

## Description
Included changes are: Promise.all for fetching all gpu pods, saToken in .kube so it is not fetched several times, storage for gpu data to prevent heavy calls being called for every user.

## How Has This Been Tested?
1. The GPU functionality should function the exact same, with less delays between calls

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by maroroman** [comment]
the user const in route-security fired a warning (unused) in my test so I commented it out.

File: backend/src/routes/api/gpu/gpuUtils.ts [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
-import * as fs from 'fs';
+
+/** Storage to prevent heavy calls from being performed for EVERY user */
+const storage: { lastFetch: number; lastValue: [boolean, number] } = {
+  lastValue: [false, 0],
+  lastFetch: 0,
+};
 
 export const getGPUNumber = async (fastify: KubeFastifyInstance): Promise<[boolean, number]> => {
+  if (storage.lastFetch >= Date.now() - 30_000) {
```
Comment (dlabaj) [inline backend/src/routes/api/gpu/gpuUtils.ts:12]: Should 30_000 be a constant?

File: backend/src/utils/route-security.ts [hunk 2]
```diff
@@ -50,7 +50,7 @@ const requestSecurityGuardNotebook = async (
   request: OauthFastifyRequest,
   username: string,
 ): Promise<void> => {
-  const user = await getUserName(fastify, request);
+  //const user = await getUserName(fastify, request);
```
Comment (dlabaj) [inline backend/src/utils/route-security.ts:53]: Can you remove this dead code.

**Review by dlabaj — COMMENTED** [review]
Few comments otherwise this looks good.

**Comment by dlabaj** [comment]
@maroroman There's merge conflicts with this issue.

File: backend/src/utils/route-security.ts (hunk 2 above)
Comment (andrewballantyne) [reply backend/src/utils/route-security.ts:53]: Don't remove this 🤔 I think we need it to validate the access of the x-forwarded token. cc @lucferbux

File: backend/src/routes/api/gpu/gpuUtils.ts (hunk 1 above)
Comment (andrewballantyne) [reply backend/src/routes/api/gpu/gpuUtils.ts:12]: Ah probably, I did this -- I was intended to slow down the costly effort of determining GPUs.

File: backend/src/utils/route-security.ts (hunk 2 above)
Comment (maroroman) [reply backend/src/utils/route-security.ts]: This got fixed after a rebase it seems

File: backend/src/routes/api/gpu/gpuUtils.ts (hunk 1 above)
Comment (maroroman) [reply backend/src/routes/api/gpu/gpuUtils.ts:12]: so just `const delay = 30000` ?

Comment (andrewballantyne) [reply backend/src/routes/api/gpu/gpuUtils.ts:12]: Might be better if we named it something like `CACHE_DELAY` and always please keep the `_` when dealing with 5+ digits in numbers.

**Pull request closed without merging**